packages/backend/server/src/plugins/oauth/controller.ts

```diff
@@ -85,65 +92,124 @@
     @Body('state') stateStr?: string,
     @Body('client_nonce') clientNonce?: string
   ) {
-    if (!code) {
-      throw new MissingOauthQueryParameter({ name: 'code' });
-    }
+    // TODO(@forehalo): refactor and remove deprecated code in 0.23
+    try {
+      if (!code) {
+        throw new MissingOauthQueryParameter({ name: 'code' });
+      }
 
-    if (!stateStr) {
-      throw new MissingOauthQueryParameter({ name: 'state' });
-    }
+      if (!stateStr) {
+        throw new MissingOauthQueryParameter({ name: 'state' });
+      }
 
-    if (typeof stateStr !== 'string' || !this.oauth.isValidState(stateStr)) {
-      throw new InvalidOauthCallbackState();
-    }
+      let isApplePost = false;
+      if (stateStr.length !== 36) {
+        try {
+          const parsed = JSON.parse(stateStr);
+          stateStr = parsed.state;
+          isApplePost = true;
```

Comment:
Whether it is enough to judge only `state.provider` and does not need `isApplePost`